File: app/src/main/java/com/example/fakeretrofitapp/MainActivity.java
package com.example.fakeretrofitapp;

import android.os.Bundle;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MainActivity extends AppCompatActivity {
    private TextView mTextViewResult;

    private JsonPlaceHolderApi jsonPlaceHolderApi;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mTextViewResult = findViewById(R.id.text_view_result);

        Gson gson = new GsonBuilder().serializeNulls().create();

        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        // Adds to Retrofit the appearance of log information after executing a request
        // such as the BODY or the HEADERS.
        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addInterceptor(loggingInterceptor)
                .build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://jsonplaceholder.typicode.com/")
                .addConverterFactory(GsonConverterFactory.create(gson))
                .client(okHttpClient)
                .build();

        jsonPlaceHolderApi = retrofit.create(JsonPlaceHolderApi.class);

        // getPosts();
        // getComments();
        // createPost();
        updatePost();
        // deletePost();
    }

    private void getPosts() {
        Map<String, String> parameters = new HashMap<>();
        parameters.put("userId", "1");
        parameters.put("_sort", "id");
        parameters.put("_order", "desc");

        Call<List<Post>> call = jsonPlaceHolderApi.getPosts(parameters);
        call.enqueue(new Callback<List<Post>>() {
            @Override
            public void onResponse(Call<List<Post>> call, Response<List<Post>> response) {
                if (!response.isSuccessful()) {
                    mTextViewResult.setText("Code: " + response.code());
                    return;
                }

                List<Post> posts = response.body();
                for (Post post : posts) {
                    String content = "";
                    content += "ID: " + post.getId() + "\n"
                            + "User ID: " + post.getUserId() + "\n"
                            + "Title: " + post.getTitle() + "\n"
                            + "Body: " + post.getBody() + "\n\n";

                    mTextViewResult.append(content);
                }

            }

            @Override
            public void onFailure(Call<List<Post>> call, Throwable t) {
                mTextViewResult.setText(t.getMessage());
            }
        });
    }

    private void getComments() {
        Call<List<Comment>> call = jsonPlaceHolderApi.getComments("posts/3/comments");
        call.enqueue(new Callback<List<Comment>>() {
            @Override
            public void onResponse(Call<List<Comment>> call, Response<List<Comment>> response) {
                if (!response.isSuccessful()) {
                    mTextViewResult.setText("Code: " + response.code());
                    return;
                }

                List<Comment> comments = response.body();
                for (Comment comment : comments) {
                    String content = "";
                    content += "Post ID: " + comment.getPostId() + "\n"
                            + "ID: " + comment.getId() + "\n"
                            + "Name: " + comment.getName() + "\n"
                            + "Email: " + comment.getEmail() + "\n"
                            + "Body: " + comment.getBody() + "\n\n";

                    mTextViewResult.append(content);
                }

            }

            @Override
            public void onFailure(Call<List<Comment>> call, Throwable t) {
                mTextViewResult.setText(t.getMessage());
            }
        });
    }

    private void createPost() {
        Post post = new Post(23, "New Title", "New Text");
        Map<String, String> fields = new HashMap<>();
        fields.put("userId", "23");
        fields.put("title", "New Title");
        fields.put("body", "New Text");

        Call<Post> call = jsonPlaceHolderApi.createPost(fields);
        call.enqueue(new Callback<Post>() {
            @Override
            public void onResponse(Call<Post> call, Response<Post> response) {
                if (!response.isSuccessful()) {
                    mTextViewResult.setText("Code: " + response.code());
                    return;
                }

                Post postResponse = response.body();
                String content = "";
                content += "Code: " + response.code() + "\n"
                        + "ID: " + postResponse.getId() + "\n"
                        + "User ID: " + postResponse.getUserId() + "\n"
                        + "Title: " + postResponse.getTitle() + "\n"
                        + "Body: " + postResponse.getBody() + "\n\n";

                mTextViewResult.setText(content);
            }

            @Override
            public void onFailure(Call<Post> call, Throwable t) {
                mTextViewResult.setText(t.getMessage());
            }
        });
    }

    private void updatePost() {
        Post post = new Post(12, null, "New Text");
        Call<Post> call = jsonPlaceHolderApi.putPost(5, post);

        call.enqueue(new Callback<Post>() {
            @Override
            public void onResponse(Call<Post> call, Response<Post> response) {
                if (!response.isSuccessful()) {
                    mTextViewResult.setText("Code: " + response.code());
                    return;
                }

                Post postResponse = response.body();
                String content = "";
                content += "Code: " + response.code() + "\n"
                        + "ID: " + postResponse.getId() + "\n"
                        + "User ID: " + postResponse.getUserId() + "\n"
                        + "Title: " + postResponse.getTitle() + "\n"
                        + "Body: " + postResponse.getBody() + "\n\n";

                mTextViewResult.setText(content);
            }

            @Override
            public void onFailure(Call<Post> call, Throwable t) {
                mTextViewResult.setText(t.getMessage());
            }
        });
    }

    private void deletePost() {
        Call<Void> call = jsonPlaceHolderApi.deletePost(5);
        call.enqueue(new Callback<Void>() {
            @Override
            public void onResponse(Call<Void> call, Response<Void> response) {
                mTextViewResult.setText("Code: " + response.code());
            }

            @Override
            public void onFailure(Call<Void> call, Throwable t) {
                mTextViewResult.setText(t.getMessage());
            }
        });
    }
}

File: app/src/main/java/com/example/fakeretrofitapp/MainActivity.kt
package com.example.fakeretrofitapp

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.example.fakeretrofitapp.models.Comment
import com.example.fakeretrofitapp.models.Post
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity() {
    private val repository: Repository = Repository()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // getPosts()
        // getComments()
        // createPost()
        // updatePost()
        deletePost()
    }

    private fun getPosts() {
        val parameters = HashMap<String, String>()
        parameters["userId"] = "1"
        parameters["_sort"] = "id"
        parameters["_order"] = "desc"

        //val call = repository.client.getPosts(1, "Title", "Body")
        val call = repository.client.getPosts(parameters)
        call.enqueue(object : Callback<List<Post>> {
            override fun onResponse(call: Call<List<Post>>, response: Response<List<Post>>) {
                if (!response.isSuccessful) {
                    text_view_result.text = "Code: ${response.code()}"
                    return
                }
                val posts = response.body()
                posts?.let { postList ->
                    for (post in postList) {
                        val content = "ID: ${post.id} User ID: ${post.userId} Title: ${post.title} Body: ${post.body}"
                        text_view_result.append(content)
                    }
                }
            }

            override fun onFailure(call: Call<List<Post>>, t: Throwable) {
                text_view_result.text = t.message
            }
        })
    }

    private fun getComments() {
        val call = repository.client.getComments("posts/3/comments")
        call.enqueue(object : Callback<List<Comment>> {
            override fun onResponse(call: Call<List<Comment>>, response: Response<List<Comment>>) {
                if (!response.isSuccessful) {
                    text_view_result.text = "Code: ${response.code()}"
                    return
                }
                val comments = response.body()
                comments?.let { commentList ->
                    for (comment in commentList) {
                        val content = "Post ID: ${comment.postId} ID: ${comment.id} Name: ${comment.name} Email: ${comment.email} Body: ${comment.body}"
                        text_view_result.append(content)
                    }
                }
            }

            override fun onFailure(call: Call<List<Comment>>, t: Throwable) {
                text_view_result.text = t.message
            }
        })
    }

    private fun createPost() {
        val fields = HashMap<String, String>()
        fields["userId"] = "23"
        fields["title"] = "New Title"
        fields["body"] = "New Text"

        val call = repository.client.createPost(fields)
        call.enqueue(object : Callback<Post?> {
            override fun onResponse(call: Call<Post?>, response: Response<Post?>) {
                if (!response.isSuccessful) {
                    text_view_result.text = "Code: ${response.code()}"
                    return
                }
                val postResponse = response.body()
                postResponse?.let {
                    val content = "Code: ${response.code()} ID: ${postResponse.id} User ID: ${postResponse.userId} Title: ${postResponse.title} Body: ${postResponse.body}"
                    text_view_result.text = content
                }
            }

            override fun onFailure(call: Call<Post?>, t: Throwable) {
                text_view_result.text = t.message
            }
        })
    }

    private fun updatePost() {
        val post = Post(12, null, "New Text")
        val headers = HashMap<String, String>()
        headers["Map-Header1"] = "def"
        headers["Map-Header2"] = "ghi"
        val call = repository.client.patchPost(headers, 5, post)
        call.enqueue(object : Callback<Post?> {
            override fun onResponse(call: Call<Post?>, response: Response<Post?>) {
                if (!response.isSuccessful) {
                    text_view_result.text = "Code: ${response.code()}"
                    return
                }
                val postResponse = response.body()
                var content = ""
                content += """
                    Code: ${response.code()}
                    ID: ${postResponse?.id}
                    User ID: ${postResponse?.userId}
                    Title: ${postResponse?.title}
                    Body: ${postResponse?.body}


                    """.trimIndent()
                text_view_result.text = content
            }

            override fun onFailure(call: Call<Post?>, t: Throwable) {
                text_view_result.text = t.message
            }
        })
    }

    private fun deletePost() {
        val call = repository.client.deletePost(5)
        call.enqueue(object : Callback<Void?> {
            override fun onResponse(call: Call<Void?>, response: Response<Void?>) {
                text_view_result.text = "Code: ${response.code()}"
            }

            override fun onFailure(call: Call<Void?>, t: Throwable) {
                text_view_result.text = t.message
            }
        })
    }
}